File: Kern/Sources/DuoKern/Box/Statistics.swift
import Foundation

/// Aggregates for the Fortschritt/Box tabs, scoped to the current direction.
public struct BoxStatistics: Sendable, Equatable {
    /// Scheduled, non-suspended cards in the current direction.
    public var activeCount: Int
    public var dueCount: Int
    public var suspendedCount: Int
    /// New-card introductions still possible today; 0 if the health gate is closed.
    public var newBudgetRemainingToday: Int
    /// Consecutive days with reviews > 0; one missed day is forgiven.
    public var streak: Int
    /// Mean FSRS retrievability over active review-phase cards; nil if none.
    public var averageRetrievability: Double?
    public var areas: [AreaStatistics]

    public init(activeCount: Int, dueCount: Int, suspendedCount: Int,
                newBudgetRemainingToday: Int, streak: Int,
                averageRetrievability: Double?, areas: [AreaStatistics]) {
        self.activeCount = activeCount
        self.dueCount = dueCount
        self.suspendedCount = suspendedCount
        self.newBudgetRemainingToday = newBudgetRemainingToday
        self.streak = streak
        self.averageRetrievability = averageRetrievability
        self.areas = areas
    }
}

public struct AreaStatistics: Sendable, Equatable {
    public var name: String
    /// Cards in the area (any status).
    public var total: Int
    /// Scheduled, non-suspended in the current direction.
    public var active: Int
    /// Review phase with stability ≥ phraseUnlockStability ("sitting" material).
    public var sitting: Int
    /// Component phrases still waiting for their components to stabilize.
    public var phrasesLocked: Int
    /// Phrases already introduced, component-free, or with all components stable.
    public var phrasesUnlocked: Int

    public init(name: String, total: Int, active: Int, sitting: Int,
                phrasesLocked: Int, phrasesUnlocked: Int) {
        self.name = name
        self.total = total
        self.active = active
        self.sitting = sitting
        self.phrasesLocked = phrasesLocked
        self.phrasesUnlocked = phrasesUnlocked
    }
}

extension BoxEngine {

    public static func statistics(state: BoxState, now: Date, calendar: Calendar) -> BoxStatistics {
        let active = activeSchedulings(state)
        let dueCount = active.filter { ($0.due ?? .distantFuture) <= now }.count
        let suspendedCount = directionSchedulings(state).filter(\.suspended).count

        return BoxStatistics(
            activeCount: active.count,
            dueCount: dueCount,
            suspendedCount: suspendedCount,
            newBudgetRemainingToday: gatedNewBudget(state, now: now, calendar: calendar),
            streak: streak(state.dailyStats, now: now, calendar: calendar),
            averageRetrievability: averageRetrievability(active, config: state.config, now: now),
            areas: areaStatistics(state)
        )
    }

    /// Mean retrievability over active review-phase cards, elapsed measured
    /// from each card's last review (last log entry date).
    static func averageRetrievability(_ active: [CardScheduling], config: BoxConfig, now: Date) -> Double? {
        let fsrs = FSRS(parameters: FSRSParameters(desiredRetention: config.desiredRetention))
        let reviewPhase = active.filter { $0.phase == .review && $0.memory != nil }
        guard !reviewPhase.isEmpty else { return nil }
        let sum = reviewPhase.reduce(0.0) { acc, sched in
            let last = sched.log.last?.date ?? sched.addedAt
            let elapsed = max(0, now.timeIntervalSince(last) / 86_400)
            return acc + fsrs.retrievability(state: sched.memory ?? MemoryState(stability: 0.1, difficulty: 5), elapsedDays: elapsed)
        }
        return sum / Double(reviewPhase.count)
    }

    /// Walk back from today. Today without reviews neither breaks the streak
    /// nor consumes forgiveness (the day isn't over); after that, exactly ONE
    /// missed day is forgiven, the next miss ends the streak.
    static func streak(_ dailyStats: [String: DayStats], now: Date, calendar: Calendar) -> Int {
        var count = 0
        var forgivenessLeft = 1
        var day = calendar.startOfDay(for: now)
        var isToday = true
        while true {
            let reviews = dailyStats[dayKey(for: day, calendar: calendar)]?.reviews ?? 0
            if reviews > 0 {
                count += 1
            } else if !isToday {
                if forgivenessLeft > 0 {
                    forgivenessLeft -= 1
                } else {
                    break
                }
            }
            isToday = false
            guard let previous = calendar.date(byAdding: .day, value: -1, to: day) else { break }
            day = previous
        }
        return count
    }

    static func areaStatistics(_ state: BoxState) -> [AreaStatistics] {
        let byArea = Dictionary(grouping: state.cards.values, by: \.area)
        return byArea.keys.sorted().map { area in
            let cards = byArea[area] ?? []
            var active = 0, sitting = 0, locked = 0, unlocked = 0
            for card in cards {
                let sched = scheduling(state, card.id)
                if let sched, !sched.suspended {
                    active += 1
                    if sched.phase == .review, let memory = sched.memory,
                       memory.stability >= state.config.phraseUnlockStability {
                        sitting += 1
                    }
                }
                if card.kind == .phrase {
                    if sched != nil || card.componentIDs.isEmpty || isPhraseUnlocked(state, card) {
                        unlocked += 1
                    } else {
                        locked += 1
                    }
                }
            }
            return AreaStatistics(name: area, total: cards.count, active: active,
                                  sitting: sitting, phrasesLocked: locked, phrasesUnlocked: unlocked)
        }
    }
}

File: kmp/kern/src/commonMain/kotlin/net/spross/kern/box/Statistics.kt
package net.spross.kern.box

import kotlin.math.max
import kotlin.time.DurationUnit
import kotlin.time.Instant
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.minus
import net.spross.kern.fsrs.Fsrs
import net.spross.kern.model.CardKind
import net.spross.kern.model.CardPhase
import net.spross.kern.model.DayStats
import net.spross.kern.model.UnitKey
import net.spross.kern.model.UnitScheduling

/** Aggregates for progress UI. All headline counts are CONCEPT-denominated. */
data class BoxStatistics(
    /** Concepts with at least one active (scheduled, non-suspended) unit. */
    val activeCount: Int,
    /** Concepts with at least one active unit due now. */
    val dueCount: Int,
    /** Concepts whose scheduled units are ALL suspended (fully out of rotation). */
    val suspendedCount: Int,
    /** Concepts that could enter the pool now; 0 when the health gate is closed. */
    val newSlotsAvailable: Int,
    /** Consecutive days with reviews > 0; one missed day is forgiven. */
    val streak: Int,
    /** Mean FSRS retrievability over active review-phase UNITS; null if none. */
    val averageRetrievability: Double?,
    val areas: List<AreaStatistics>,
)

data class AreaStatistics(
    val name: String,
    /** Concepts in the area (any status). */
    val total: Int,
    /** Concepts with at least one active unit. */
    val active: Int,
    /** Concepts whose produce unit sits in Review with stability >= unlock threshold. */
    val sitting: Int,
    /** Component phrases still waiting for their components to stabilize. */
    val phrasesLocked: Int,
    /** Phrases already introduced, component-free, or with all components stable. */
    val phrasesUnlocked: Int,
)

internal object Statistics {

    fun statistics(state: BoxState, nowEpochMillis: Long, tzId: String): BoxStatistics {
        val now = Instant.fromEpochMilliseconds(nowEpochMillis)
        val active = Inventory.active(state)
        val suspendedConcepts = Inventory.scheduled(state)
            .groupBy { it.cardId }
            .count { (_, units) -> units.all { it.suspended } }
        return BoxStatistics(
            activeCount = active.mapTo(mutableSetOf()) { it.cardId }.size,
            dueCount = active
                .filter { it.due != null && it.due <= now }
                .mapTo(mutableSetOf()) { it.cardId }
                .size,
            suspendedCount = suspendedConcepts,
            newSlotsAvailable = Growth.gatedNewBudget(state, nowEpochMillis),
            streak = streak(state.dailyStats, nowEpochMillis, tzId),
            averageRetrievability = averageRetrievability(state, active, now),
            areas = areaStatistics(state),
        )
    }

    /**
     * Walk back from today. Today without reviews neither breaks the streak nor consumes
     * forgiveness (the day isn't over); afterwards exactly ONE 0-review day is forgiven,
     * the next miss ends the streak. Forgiven days do not increment the count.
     */
    fun streak(dailyStats: Map<String, DayStats>, nowEpochMillis: Long, tzId: String): Int {
        var count = 0
        var forgivenessLeft = 1
        var day = localDate(nowEpochMillis, tzId)
        var isToday = true
        while (true) {
            val reviews = dailyStats[day.toString()]?.reviews ?: 0
            if (reviews > 0) {
                count += 1
            } else if (!isToday) {
                if (forgivenessLeft > 0) forgivenessLeft -= 1 else break
            }
            isToday = false
            day = day.minus(1, DateTimeUnit.DAY)
        }
        return count
    }

    /** Elapsed measured from each unit's last review (last log entry date). */
    private fun averageRetrievability(
        state: BoxState,
        active: List<UnitScheduling>,
        now: Instant,
    ): Double? {
        val review = active.filter { it.phase == CardPhase.Review && it.memory != null }
        if (review.isEmpty()) return null
        val fsrs = Fsrs(state.config.fsrsParameters())
        val sum = review.sumOf { sched ->
            val last = sched.log.lastOrNull()?.date ?: sched.addedAt
            val elapsed = max(0.0, (now - last).toDouble(DurationUnit.DAYS))
            fsrs.retrievability(elapsed, sched.memory!!.stability)
        }
        return sum / review.size
    }

    private fun areaStatistics(state: BoxState): List<AreaStatistics> {
        val activeCards = Inventory.active(state).mapTo(mutableSetOf()) { it.cardId }
        return state.cards.values.groupBy { it.area }.entries
            .sortedBy { it.key }
            .map { (area, cards) ->
                var active = 0
                var sitting = 0
                var locked = 0
                var unlocked = 0
                for (card in cards) {
                    if (card.id in activeCards) active += 1
                    val produce = state.scheduling[UnitKey.produce(card.id).encoded]
                    if (produce != null && !produce.suspended &&
                        produce.phase == CardPhase.Review &&
                        (produce.memory?.stability ?: 0.0) >= state.config.phraseUnlockStability
                    ) {
                        sitting += 1
                    }
                    if (card.kind == CardKind.Phrase) {
                        val open = produce != null || card.components.isEmpty() ||
                            Growth.isPhraseUnlocked(state, card)
                        if (open) unlocked += 1 else locked += 1
                    }
                }
                AreaStatistics(area, cards.size, active, sitting, locked, unlocked)
            }
    }
}
